### PR TITLE
feat(semantics): Phase 4 - LSP Migration Infrastructure

### PR DESCRIPTION
**Pull request by albertocavalcante**
# Phase 4: LSP Migration Infrastructure

## Summary

This PR completes the infrastructure phase for migrating LSP providers to use the new `GroovySemantics` API from `semantics-native`. The new `SemanticTypeResolver` bridge class is production-ready and fully tested.

### 1. Fixed Gradle Module Collision
- **Problem**: Both `:parser:native` and `:semantics:native` had the same leaf name `native`, causing Gradle to substitute one for the other during dependency resolution
- **Solution**: Registered `semantics-native` as a top-level project with `projectDir` mapping to `semantics/native`
- **Files**: `settings.gradle.kts`

### 2. Added Dependencies
- Added `semantics:core` and `semantics-native` to `groovy-lsp`
- **Files**: `groovy-lsp/build.gradle.kts`

### 3. Created SemanticTypeResolver Bridge
- New bridge class connecting `GroovySemantics` with LSP providers
- Methods:
  - `resolveType(node, moduleNode): SemanticType` - Resolves type using GroovySemantics
  - `toClassNode(type, moduleNode): ClassNode?` - Converts to Groovy AST ClassNode for backward compatibility
  - `formatSemanticType(type): String` - Formats for UI display
- **Files**: `groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolver.kt`

### 4. Comprehensive Test Suite
- 18 tests covering all `SemanticType` variants
- Tests for ClassNode conversion, type formatting, primitives, unions, arrays, edge cases
- **All tests passing** ✅
- **Files**: `groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolverTest.kt`

### 5. Deprecated Legacy Implementations
- Added `@Deprecated` annotations to `GroovyTypeResolver` and `TypeInferencer`
- Includes `ReplaceWith` suggestions pointing to new implementations
- **17 deprecation warnings** now guide future migration efforts
- **Files**: 
  - `groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/types/GroovyTypeResolver.kt`
  - `parser/native/src/main/kotlin/com/github/albertocavalcante/groovyparser/ast/TypeInferencer.kt`

## Testing

```bash
# All SemanticTypeResolver tests pass
./gradlew :groovy-lsp:test --tests SemanticTypeResolverTest
# Result: 18/18 tests passing ✅

# No regressions
make test
# Result: All tests passing ✅

# Deprecation warnings visible
./gradlew :groovy-lsp:compileKotlin
# Result: 17 deprecation warnings as expected ✅
```

## Type of Change

- [x] New feature (infrastructure for future LSP provider migration)
- [x] Refactoring (deprecation of legacy implementations)

## Migration Strategy

This PR takes a **conservative approach**:
- ✅ Infrastructure is ready and tested
- ✅ Legacy code is deprecated with clear migration paths
- ✅ **Zero risk** of breaking existing functionality
- ✅ Providers can migrate incrementally in future PRs

The deprecation warnings guide future work without forcing immediate changes.

## Checklist

- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes (18 tests)
- [x] All new and existing tests passed
- [x] Code is properly formatted (`spotlessApply`)
- [x] No regressions in existing functionality

## Related Issues

Part of the semantic layer implementation work.

## Next Steps (Future PRs)

1. Migrate `InlayHintsProvider` to use `SemanticTypeResolver` for better type hints
2. Gradually replace `TypeInferencer` calls with `SemanticTypeResolver`
3. Remove deprecated implementations after full migration

---

> [!NOTE]
> Implements semantic-aware LSP and tooling upgrades, plus a major PR review CLI refactor.
> 
> - LSP: Introduces `SemanticTypeResolver` and wires it through providers (completion, signature help, inlay hints) and native engine; replaces legacy hover with `HoverContentGenerator`; removes `HoverNodeConverters`; exposes `ClasspathService.getTypeSolver`; adds `semantics:core` and `semantics-native` deps
> - Completions/Signature Help/Inlay Hints: replace `TypeInferencer` usage with semantic resolution; adjust context detection to use semantic types; update tests accordingly
> - Hover: new semantic-backed generator and documentation merge in `HoverProvider`
> - VS Code tools: prefer `*-all.jar` when selecting server; validate JAR by size (>5MB) and ZIP contents (e.g., `ClassNode.class`); minor monorepo local-build logic tweaks
> - Build/Run: Makefile switches run targets to `gls-*-all.jar`
> - CLI (`.agent/scripts/pr.py`): adds reviewer-specific parsers, severity sorting, human/LLM/JSON formatters, output mode auto-detection, `--compact` option, and a `reject` command
> - Misc: add Python `__pycache__/` to `.gitignore`; small test/runtime type fixes
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c337b685ec570922f745f309cd59f2c7e3610f31. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Semantic-aware hovers, inlay hints and completions with richer, type-aware formatting; runtime type-solver exposed.

* **Deprecations**
  * Legacy type-inference APIs now marked deprecated with migration hints toward the new semantic APIs.

* **Chores**
  * Native semantics module wired into the build; hover pipeline replaced with semantic content generator; artifact naming and server validation/selection improved; CLI review tool enhanced.

* **Tests**
  * New and expanded tests for semantic type mapping, hover content, inlay hints and completions.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: parser/core/src/main/kotlin/com/github/albertocavalcante/groovyparser/resolution/typesolvers/ReflectionTypeSolver.kt [hunk 1]
```diff
@@ -17,11 +17,13 @@ import java.util.concurrent.ConcurrentHashMap
  *
  * @param jreOnly If true, only resolve types in java.* and javax.* packages
  */
-class ReflectionTypeSolver(private val jreOnly: Boolean = true) : TypeSolver {
+class ReflectionTypeSolver(
+    private val classLoader: ClassLoader = ReflectionTypeSolver::class.java.classLoader,
+    private val jreOnly: Boolean = true,
+) : TypeSolver {
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply parser/core/src/main/kotlin/com/github/albertocavalcante/groovyparser/resolution/typesolvers/ReflectionTypeSolver.kt]: Fixed in fbaef420. Reverted parameter order to avoid breaking compatibility.

File: groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolverTest.kt [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    @Test
+    fun `formats nested Array type correctly`() {
+        val type =
+            SemanticType.Array(
+                SemanticType.Array(SemanticType.Primitive(PrimitiveKind.INT)),
+            )
+        val formatted = resolver.formatSemanticType(type)
+
+        assertThat(formatted).isEqualTo("int[][]")
+    }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolverTest.kt]: Fixed in fbaef420. Added test coverage for resolveType.

File: groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolver.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    /**
+     * Format SemanticType for display in hover, inlay hints, etc.
+     */
+    fun formatSemanticType(type: SemanticType): String = when (type) {
+        is SemanticType.Known -> type.fqn.substringAfterLast('.')
+        is SemanticType.Primitive -> type.kind.name.lowercase()
+        is SemanticType.Dynamic -> type.hint ?: "def"
+        is SemanticType.Unknown -> "unresolved"
+        is SemanticType.Union -> type.types.joinToString(" | ") { formatSemanticType(it) }
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolver.kt]: Fixed in 64d92333. Enforced sorted order for union types.

File: groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/providers/inlayhints/InlayHintsProviderTest.kt [hunk 4]
```diff
@@ -281,6 +305,10 @@ class InlayHintsProviderTest {
                 null,
             )
 
+            mockResolverType("path", "String")
+
+            mockResolverType("path", "String")
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/test/kotlin/com/github/albertocavalcante/groovylsp/providers/inlayhints/InlayHintsProviderTest.kt]: Fixed in 793f1721. Removed duplicate mockResolverType call.

File: groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolver.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply groovy-lsp/src/main/kotlin/com/github/albertocavalcante/groovylsp/types/SemanticTypeResolver.kt]: Fixed in 793f1721. Added sortedBy to Union type formatting for deterministic output.

File: tests/e2e/kotlin/com/github/albertocavalcante/groovylsp/e2e/StepExecutors.kt [hunk 6]
```diff
@@ -145,9 +145,7 @@ class InitializeStepExecutor : StepExecutor<ScenarioStep.Initialize> {
         val params = InitializeParams().apply {
             processId = ProcessHandle.current().pid().toInt()
             val finalRootUri = step.rootUri ?: context.workspace.rootUri
-            if (finalRootUri != null) {
-                workspaceFolders = listOf(WorkspaceFolder(finalRootUri, context.workspace.rootDir.name))
-            }
+            workspaceFolders = listOf(WorkspaceFolder(finalRootUri, context.workspace.rootDir.name))
```
[the remark this answers is not in the document]
Comment (albertocavalcante) [reply tests/e2e/kotlin/com/github/albertocavalcante/groovylsp/e2e/StepExecutors.kt]: This is a false positive. The `context.workspace.rootUri` is always populated from the test fixture setup during E2E test initialization. The Initialize step is always preceded by workspace setup that ensures rootUri is set. A null value here would indicate a test setup bug, not a runtime issue that needs defensive coding.
